### PR TITLE
config-tools: do not exit when the board inspector runs in hypervisor

### PR DESCRIPTION
**Pull request by yuchuyang**
While running in a nested environment, such as qemu, parse the board information should be allowed even it is not in a native environment.
Replace the error with warning message and does not exit the program.

Tracked-On: #6208
Signed-off-by: Yang,Yu-chu <yu-chu.yang@intel.com>

### CODE REVIEW

**Comment by yuchuyang** [comment]
@wenlingz  Please hold this PR, do not merge yet. Thanks.

**Comment by NanlinXie** [comment]
@fuzhongl , could you help to verify this?

File: misc/config_tools/board_inspector/cli.py
Comment (dbkinder) [inline misc/config_tools/board_inspector/cli.py:24]: We avoid saying "please".  Also you might include saying it looks like you're running the tool in a VM.  Perhaps:
"The board inspector must be run in a native OS environment and not from a VM."

File: misc/config_tools/board_inspector/cli.py
Comment (dbkinder) [inline misc/config_tools/board_inspector/cli.py:29]:
```suggestion
        # Check if this is native os
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
Does  ``--no-native-check`` allow the tool to work in a VM? Does the tool still work as expected then?  Why do we need this option?

Also, we have no documentation about the board inspector options :(

**Comment by yuchuyang** [comment]
@dbkinder This option by far is particular for qemu (it maybe for other nested env in the future) like the commit message stats "While running in a nested environment, such as qemu, parse the board information should be allowed even it is not in a native environment."

qemu may configure the env unlike how it does without qemu. It is needed to get the board information which is configured by qemu, that's why this flag is implemented.

**Comment by fuzhongl** [comment]
> @fuzhongl , could you help to verify this?
Verified! Please check following log:

```
root@ubuntu:/home/acrn/acrn-hypervisor# python3 misc/config_tools/board_inspector/cli.py --no-native-check qemu
Warning:The Cx data for ACRN relies on acpi_idle driver but it is not found, Warning:please make sure ACPI Cstate is enabled in BIOS.
Warning:No scaling_driver found.
Resource Allocation is not supported!
qemu.xml has been generated successfully!
root@ubuntu:/home/acrn/acrn-hypervisor# ls misc/config_tools/board_inspector
acpiparser  cli.py  cpuparser  extractors  legacy  lib  memmapparser  pcieparser  qemu.xml  README  smbiosparser
```

**Comment by gvancuts** [comment]
How about we do this the other way around, i.e.:
1. Check whether we are running in KVM (Python equivalent of `cpuid -l  0x40000000 | grep -c KVMKVMVKM`
2. If we are, allow the tool to run but display a warning to the user that they are creating a $(BOARD).xml for a virtual machine

That way we do not need a new special flag.

**Comment by yuchuyang** [comment]
@gvancuts tool may not run inside a VM, for example, virtual box would fail. This flag is recommended only if user knows what exactly she/he is doing.
I would think if we shall explicitly specify the flag is for qemu only(by far) to prevent out-of-box usage...

**Comment by gvancuts** [comment]
> @gvancuts tool may not run inside a VM, for example, virtual box would fail. This flag is recommended only if user knows what exactly she/he is doing.
> I would think if we shall explicitly specify the flag is for qemu only(by far) to prevent out-of-box usage...

Yes, this is my proposal. Check for `cpuid -l 0x40000000 | grep -c KVMKVMVKM` to be equal or greater than 1 and if so, it means we are running in QEMU/KVM. In that case, allow the script to proceed **and** display a warning so that the user is aware of the special case.

**Comment by yuchuyang** [comment]
> > @gvancuts tool may not run inside a VM, for example, virtual box would fail. This flag is recommended only if user knows what exactly she/he is doing.
> > I would think if we shall explicitly specify the flag is for qemu only(by far) to prevent out-of-box usage...
> 
> Yes, this is my proposal. Check for `cpuid -l 0x40000000 | grep -c KVMKVMVKM` to be equal or greater than 1 and if so, it means we are running in QEMU/KVM. In that case, allow the script to proceed **and** display a warning so that the user is aware of the special case.

There are 2 ways to remove the "--no-native-check" flag: 
1. if we don't care about which hypervisor is now, based on this patch, we could remove the --no-native-check flag and replace the logging.error(f"The board inspector must be run in a native OS environment and not from a VM.")
 with logging.warning(f"The board inspector is running on top of hypervisor, you should running this on top of QEMU only.") and do not exit. If it fails, it gets sufficient information.
2. as Geoffroy proposed, If we want to be precise, cpuid with 0x40000000 would work too.. just extra work to implement a class here. Btw, @gvancuts do you know if in the 0x40000000 would return "" KVMKVMKVM  " or "TCGTCGTCGTCG"? 

@junjiemao1 any comments?

**Comment by gvancuts** [comment]
>     2. as Geoffroy proposed, If we want to be precise, cpuid with 0x40000000 would work too.. just extra work to implement a class here. Btw, @gvancuts do you know if in QEMU, cpuid with 0x40000000 would return "" KVMKVMKVM  " or "TCGTCGTCGTCG"?

It will return `KVMKVMKVM`.

Out of curiosity, where did you get this `TCGTCGTCGTCG` from? I have never seen it, I have seen the CPUID for ACRN, Xen, Jailhouse and QEMU but not this one.

**Comment by junjiemao1** [comment]
> > > @gvancuts tool may not run inside a VM, for example, virtual box would fail. This flag is recommended only if user knows what exactly she/he is doing.
> > > I would think if we shall explicitly specify the flag is for qemu only(by far) to prevent out-of-box usage...
> > 
> > 
> > Yes, this is my proposal. Check for `cpuid -l 0x40000000 | grep -c KVMKVMVKM` to be equal or greater than 1 and if so, it means we are running in QEMU/KVM. In that case, allow the script to proceed **and** display a warning so that the user is aware of the special case.
> 
> There are 2 ways to remove the "--no-native-check" flag:
> 
> 1. if we don't care about which hypervisor is now, based on this patch, we could remove the --no-native-check flag and replace the logging.error(f"The board inspector must be run in a native OS environment and not from a VM.")
>    with logging.warning(f"The board inspector is running on top of hypervisor, you should running this on top of QEMU only.") and do not exit. If it fails, it gets sufficient information.
> 2. as Geoffroy proposed, If we want to be precise, cpuid with 0x40000000 would work too.. just extra work to implement a class here. Btw, @gvancuts do you know if in QEMU, cpuid with 0x40000000 would return "" KVMKVMKVM  " or "TCGTCGTCGTCG"?
> 
> @junjiemao1 any comments?

It depends on our expected usage. If KVM (with nested support) is the only target, we can check `` KVMKVMKVM ``. ``TCGTCGTCGTCG`` is the leaf reported by Qemu with KVM disabled (where ``TCG``, i.e. ``Tiny Code Generator``, is the intermediate representation of binary in Qemu for binary translation).

**Comment by yuchuyang** [comment]
@gvancuts  I have removed the flag and replace the error message with warning instead. Now we only support few nested environment, but could be expand in the future. I would like keep it simple, no flag, no exit but a warning message and do not block running board inspector in any condition.

**Comment by wenlingz** [comment]
@yuchuyang please notify me once this PR can be merged, thanks

**Comment by gvancuts** [comment]
> @gvancuts I have removed the flag and replace the error message with warning instead.

Thanks.

> Now we only support few nested environment, but could be expand in the future.

Which ones do we support (beyond KVM)?

> I would like keep it simple, no flag, no exit but a warning message and do not block running board inspector in any condition.

Yes, this is a good approach. But we need to catch the fact we're running in KVM since this is supported, and hence no "unexpected results" should occur.

**Comment by yuchuyang** [comment]
> > @gvancuts I have removed the flag and replace the error message with warning instead.
> 
> Thanks.
> 
> > Now we only support few nested environment, but could be expand in the future.
> 
> Which ones do we support (beyond KVM)?
> 
Not for now except the KVM. I keep the positive thought :)

> > I would like keep it simple, no flag, no exit but a warning message and do not block running board inspector in any condition.
> 
> Yes, this is a good approach. But we need to catch the fact we're running in KVM since this is supported, and hence no "unexpected results" should occur.

KVM is widely used for different hypervisor, for Acrn + qemu/KVM case, it's sufficient but may not always true since qemu/KVM is one of the possibility. To ensure that acrn can work with the board.xml, it should based on its capability, for example: vmx. Based on this statement, I would think do the schema validation while compiling may be sufficient.

**Comment by yuchuyang** [comment]
@wenlingz  Please helps to merge this PR, thanks.

**Comment by wenlingz** [comment]
@NanlinXie please help approve this PR, thanks

**Comment by NanlinXie** [comment]
@gvancuts , are you ok for this patch?

**Comment by gvancuts** [comment]
> > > Now we only support few nested environment, but could be expand in the future.
> > 
> > 
> > Which ones do we support (beyond KVM)?
> 
> Not for now except the KVM. I keep the positive thought :)
> 
> > > I would like keep it simple, no flag, no exit but a warning message and do not block running board inspector in any condition.
> > 
> > 
> > Yes, this is a good approach. But we need to catch the fact we're running in KVM since this is supported, and hence no "unexpected results" should occur.
> 
> KVM is widely used for different hypervisor, for Acrn + qemu/KVM case, it's sufficient but may not always true since qemu/KVM is one of the possibility. To ensure that acrn can work with the board.xml, it should based on its capability, for example: vmx. Based on this statement, I would think do the schema validation while compiling may be sufficient.

I am not following you here. What I think we should do is, check if running in a native env:
* if true just generate the `$(BAORD).xml`
* if running in QEMU/KVM, print an indication the tool is collecting the `$(BOARD).xml` for a QEMU/KVM Virtual Machine (CM)
* if running in a different hypervisor, print a warning that this may lead to unpredictable results

**Comment by gvancuts** [comment]
> @gvancuts , are you ok for this patch?

Not yet

**Comment by yuchuyang** [comment]
> > KVM is widely used for different hypervisor, for Acrn + qemu/KVM case, it's sufficient but may not always true since qemu/KVM is one of the possibility. To ensure that acrn can work with the board.xml, it should based on its capability, for example: vmx. Based on this statement, I would think do the schema validation while compiling may be sufficient.
> 
> I am not following you here. What I think we should do is, check if running in a native env:
> 
> * if true just generate the `$(BAORD).xml`
> * if running in QEMU/KVM, print an indication the tool is collecting the `$(BOARD).xml` for a QEMU/KVM Virtual Machine (CM)
> * if running in a different hypervisor, print a warning that this may lead to unpredictable results

If tool run inside the virtualbox(this is not the case we will do using ACRN, but anyone can think out of box), it is running  on top of hypervisor and it is using KVM. However, it doesn't mean Acrn will work in such environment.

**Comment by gvancuts** [comment]
> 
> If tool runs inside the virtualbox(this is not the case we will do using ACRN, but anyone can think out of box), it is running on top of hypervisor and it is using KVM. However, it doesn't mean Acrn will work in such environment.

Ah, I see!! I did not know that running in VirtualBox would return that `KVMKVMKVM` string too.

I would suggest we return a warning then and state in there that only QEMU/KVM is validated.

File: misc/config_tools/board_inspector/cli.py
Comment (yuchuyang) [inline misc/config_tools/board_inspector/cli.py:26]: @gvancuts Any suggestion of improving logging message?

**Comment by pengfujx** [comment]
OK to verify doc

File: misc/config_tools/board_inspector/cli.py
Comment (gvancuts) [reply misc/config_tools/board_inspector/cli.py:26]: How about the following text:
```
Board inspector is running inside a Virtual Machine (VM). Running ACRN inside a VM is only supported under KVM/QEMU.
Unexpected results may occur when deviating from that combination.
```

**Review by gvancuts — APPROVED** [review]

**Comment by yuchuyang** [comment]
@wenlingz  This PR is ready to merge. Thanks!

**Comment by wenlingz** [comment]
@NanlinXie @terryzouhao please help approve this PR, thanks

**Comment by wenlingz** [comment]
@terryzouhao please help approve this PR, thanks

**Review by terryzouhao — APPROVED** [review]

**Pull request merged**
